streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -38,346 +40,603 @@
+
+        // the current assignment should contain the all partitions to remove
+        if (!assignment.containsAll(partitions)) {
+            throw new IllegalStateException("The current assignment " + assignment + " " +
+                "does not contain some of the partitions " + partitions + " for resuming.");
+        }
+        restoreConsumer.resume(partitions);
+    }
+
+    private void prepareChangelogs(final Map<TopicPartition, StateStoreMetadata> newPartitionsToRestore) {
+        // separate those who do not have the current offset loaded from checkpoint
+        final Set<TopicPartition> newPartitionsWithoutStartOffset = new HashSet<>();
+
+        for (final Map.Entry<TopicPartition, StateStoreMetadata> entry : newPartitionsToRestore.entrySet()) {
+            final TopicPartition partition = entry.getKey();
+            final StateStoreMetadata storeMetadata = entry.getValue();
+            final Long currentOffset = storeMetadata.offset;
+            final Long endOffset = changelogs.get(partition).restoreEndOffset;
+
+            if (currentOffset != null) {
+                restoreConsumer.seek(partition, currentOffset);
+
+                log.debug("Start restoring changelog partition {} from current offset {} to end offset {}.",
+                    partition, currentOffset, recordEndOffset(endOffset));
+            } else {
+                log.debug("Start restoring changelog partition {} from the beginning offset {} to end offset {} " +
+                    "since we cannot find current offset.", partition, recordEndOffset(endOffset));
+
+                newPartitionsWithoutStartOffset.add(entry.getKey());
+            }
+
+            // we do not trigger restore listener for standby tasks (whose end offset is unknown)
+            // TODO: maybe we can lift this restriction but need to go through a KIP since endOffset now would be null
+            if (state == ChangelogReaderState.ACTIVE_RESTORING) {
+                // NOTE if there's no offset to seek, i.e. we would seek from the beginning, we trigger callback with 0L
+                final long startOffset = currentOffset == null ? 0L : currentOffset;
+                stateRestoreListener.onRestoreStart(partition, storeMetadata.stateStore.name(), startOffset, endOffset);
             }
         }
 
-        return false;
+        // optimization: batch all seek-to-beginning offsets in a single request
+        if (!newPartitionsWithoutStartOffset.isEmpty()) {
+            restoreConsumer.seekToBeginning(newPartitionsWithoutStartOffset);
+        }
+    }
+
+    @Override
+    public void remove(final List<TopicPartition> revokedChangelogs) {
+        for (final TopicPartition partition : revokedChangelogs) {
+            ChangelogMetadata changelogMetadata = changelogs.remove(partition);
+            changelogMetadata.clear();
+        }
+
+        removeChangelogsFromRestoreConsumer(revokedChangelogs);
+    }
+
+    @Override
+    public void clear() {
```

Comment:
When the thread is being closed this should be called (already included in this PR, just FYI).